.circleci/config.yml

```diff
@@ -46,58 +36,43 @@ references:
       name: Install monitoring tools
       command: make setup-monitoring
 
+  build_images: &build_images
+    run:
+      name: Build images
+      command: make images
+
 jobs:
   build:
-    <<: *container_config
+    <<: *machine_config
     steps:
       - checkout
-      - *restore_venv
       - run:
-          name: Create python virtual environment
+          name: Build and push Docker image to DockerHub
           command: |
-            virtualenv -p python3 .venv/py3
-      - *activate_venv
-      - *install_pip
-      - save_cache:
-          key: *venv_cache_key
-          paths:
-            - ~/.venv/py3
-            - ~/.cache/pip
-      - run:
-          name: Install ansible roles
-          command: |
-            make ansible/roles
-      - save_cache:
-          key: *roles_cache_key
-          paths:
-            ansible/roles
+            docker login -u $DOCKER_USER -p $DOCKER_PASS
+            docker build -t ${DOCKERHUB_REPO:?}:${CIRCLE_BRANCH:?} .
+            docker push ${DOCKERHUB_REPO:?}:${CIRCLE_BRANCH:?}
 
   check:
-    <<: *container_config
+    <<: *machine_config
     steps:
       - checkout
-      - *restore_venv
-      - *activate_venv
-      - *install_pip
+      - *pull_current_image
```

Comment:
This seems to me another chicken-and-egg problem, and I do not see this self-healing for any branch (this check job is run for all branches but packer) unless a manual push of image for that branch is done.

@dincho Could you please elaborate?

Reply:
I see the image pulled here is used below in .circleci/docker-run.sh so this needs to be clarified.

Reply:
The image (per branch) is built in the `build` job above. The `check` job has `build` as requirement, thus `build` is always run before `check` and it's guaranteed that the image is already build.

However, the CircleCI machine scheduler may assign *the same* or different VM worker to the `check` job than the `build` job, in this case ([example workflow](https://circleci.com/workflow-run/d0e8ff39-08bf-4636-ab81-7bd54bc1e8ee)) the image is not seen by the `check` job and has to be pulled from a remote (dockerhub). 

In most of the cases it shall be already available on the `check` VM worker e.g. see the pull step in [this workflow](https://circleci.com/gh/aeternity/infrastructure/1186)

Reply:
> The image (per branch) is built in the build job above. The check job has build as requirement

ah - thanks, I had missed this.

Other issue I see is that concurrent jobs running for same branch may overlap hence the check job may pull the image built for another workflow on the same branch.

I see no job requiring build job rather than check. Any reasons for not merging the build and check jobs as one job - so to avoid the potential overlap of jobs for same branch (and reduce complexity)?

Reply:
Good idea, see 1deb63d